Mathlib/FieldTheory/IsAlgClosed/Basic.lean

```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+  natDegree_map_eq_of_injective hf ▸ roots_card_eq_natDegree
+
+theorem map_roots_card_eq_natDegree_from_simpleRing {A B : Type*} [Ring A] [IsSimpleRing A]
+    [Field B] [IsAlgClosed B] (f : A →+* B) {p : A[X]} : (p.map f).roots.card = p.natDegree :=
+  natDegree_map_from_simpleRing f p ▸ roots_card_eq_natDegree
+
+theorem aroots_card_eq_natDegree_of_leadingCoeff_ne_zero {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (hf : algebraMap A B p.leadingCoeff ≠ 0) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_leadingCoeff_ne_zero _ hf
+
+theorem aroots_card_eq_natDegree_of_isUnit_leadingCoeff {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (h : IsUnit p.leadingCoeff) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_isUnit_leadingCoeff _ h
+
+theorem aroots_card_eq_natDegree {A B : Type*} [CommRing A] [Field B] [IsAlgClosed B] [Algebra A B]
+    [FaithfulSMul A B] {p : A[X]} : (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_injective _ <| FaithfulSMul.algebraMap_injective _ _
+
+theorem dvd_iff_roots_le_roots [IsAlgClosed k] {p q : k[X]} (hp : p ≠ 0) (hq : q ≠ 0) :
+    p ∣ q ↔ p.roots ≤ q.roots :=
+  Splits.dvd_iff_roots_le_roots (splits _) hp hq
+
+theorem associated_iff_roots_eq_roots [IsAlgClosed k] {p q : k[X]} (hp : p ≠ 0) (hq : q ≠ 0) :
+    Associated p q ↔ p.roots = q.roots :=
+  ⟨roots_congr, fun h ↦ associated_of_dvd_dvd (dvd_iff_roots_le_roots hp hq |>.mpr <| le_of_eq h)
+    (dvd_iff_roots_le_roots hq hp |>.mpr <| le_of_eq h.symm)⟩
+
 theorem exists_eval₂_eq_zero_of_injective {R : Type*} [Semiring R] [IsAlgClosed k] (f : R →+* k)
     (hf : Function.Injective f) (p : R[X]) (hp : p.degree ≠ 0) : ∃ x, p.eval₂ f x = 0 :=
   let ⟨x, hx⟩ := exists_root (p.map f) (by rwa [degree_map_eq_of_injective hf])
   ⟨x, by rwa [eval₂_eq_eval_map, ← IsRoot]⟩
 
-theorem exists_eval₂_eq_zero {R : Type*} [DivisionRing R] [IsAlgClosed k] (f : R →+* k) (p : R[X])
-    (hp : p.degree ≠ 0) : ∃ x, p.eval₂ f x = 0 :=
+theorem exists_eval₂_eq_zero {R : Type*} [Ring R] [IsSimpleRing R] [IsAlgClosed k] (f : R →+* k)
+    (p : R[X]) (hp : p.degree ≠ 0) : ∃ x, p.eval₂ f x = 0 :=
   exists_eval₂_eq_zero_of_injective f f.injective p hp
 
 variable (k)
 
 theorem exists_aeval_eq_zero_of_injective {R : Type*} [CommSemiring R] [IsAlgClosed k] [Algebra R k]
     (hinj : Function.Injective (algebraMap R k)) (p : R[X]) (hp : p.degree ≠ 0) :
     ∃ x : k, aeval x p = 0 :=
-  exists_eval₂_eq_zero_of_injective (algebraMap R k) hinj p hp
+  exists_eval₂_eq_zero_of_injective _ hinj p hp
 
 theorem exists_aeval_eq_zero {R : Type*} [Field R] [IsAlgClosed k] [Algebra R k] (p : R[X])
     (hp : p.degree ≠ 0) : ∃ x : k, aeval x p = 0 :=
-  exists_eval₂_eq_zero (algebraMap R k) p hp
+  exists_eval₂_eq_zero _ p hp
+
+theorem exists_aeval_eq_zero' {R : Type*} [CommSemiring R] [IsAlgClosed k] [Algebra R k]
```

Comment:
Can you add a docstring describing the difference of this with the unprimed one? Thanks.